datafusion/physical-expr/src/expressions/case.rs

```diff
@@ -291,12 +292,67 @@ impl PhysicalExpr for CaseExpr {
 /// Create a CASE expression
 pub fn case(
     expr: Option<Arc<dyn PhysicalExpr>>,
-    when_thens: &[WhenThen],
+    when_thens: Vec<WhenThen>,
     else_expr: Option<Arc<dyn PhysicalExpr>>,
+    input_schema: &Schema,
 ) -> Result<Arc<dyn PhysicalExpr>> {
+    // all the result of then and else should be convert to a common data type,
+    // if they can be coercible to a common data type, return error.
+    let coerce_type = get_case_common_type(&when_thens, else_expr.clone(), input_schema);
+    let (when_thens, else_expr) = match coerce_type {
+        None => Err(DataFusionError::Plan(format!(
+            "Can't get a common type for then {:?} and else {:?} expression",
+            when_thens, else_expr
+        ))),
+        Some(data_type) => {
+            // cast then expr
+            let left = when_thens
+                .into_iter()
+                .map(|(when, then)| {
+                    let then = try_cast(then, input_schema, data_type.clone()).unwrap();
+                    (when, then)
+                })
+                .collect::<Vec<WhenThen>>();
+            let right = match else_expr {
+                None => None,
+                Some(expr) => Some(try_cast(expr, input_schema, data_type.clone())?),
+            };
+            Ok((left, right))
+        }
+    }?;
+
     Ok(Arc::new(CaseExpr::try_new(expr, when_thens, else_expr)?))
 }
 
+fn get_case_common_type(
+    when_thens: &[WhenThen],
+    else_expr: Option<Arc<dyn PhysicalExpr>>,
```

Comment:
```suggestion
    else_expr: Option<&PhysicalExpr>,
```

I wonder if there is a reason it needs to get a copy, or would a reference work too?

Reply:
The reference is not work for this.